frame/system/src/migrations/mod.rs

```diff
@@ -0,0 +1,182 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Migrate the reference counting state.
+
+use codec::{Decode, Encode, FullCodec};
+use frame_support::{
+	pallet_prelude::ValueQuery, traits::PalletInfoAccess, weights::Weight, Blake2_128Concat,
+	RuntimeDebug,
+};
+use sp_std::prelude::*;
+
+/// Type used to encode the number of references an account has.
+type RefCount = u32;
+
+/// All balance information for an account.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Default, RuntimeDebug, MaxEncodedLen)]
+struct AccountData<Balance> {
+	free: Balance,
+	reserved: Balance,
+	misc_frozen: Balance,
+	fee_frozen: Balance,
+}
+
+/// Information of an account.
+#[derive(Clone, Eq, PartialEq, Default, RuntimeDebug, Encode, Decode)]
+struct AccountInfo<Index, AccountData> {
+	nonce: Index,
+	consumers: RefCount,
+	providers: RefCount,
+	sufficients: RefCount,
+	data: AccountData,
+}
+
+/// Trait to implement to give information about types used for migration
+pub trait V2ToV3 {
```

Comment:
This migration (triple reference counting) is required for substrate V3. So I would implement it through a custom migration by implementing the trait V2ToV3 that binds the relevant runtime quantities (Index, Balances etc.) to the migration logic. I have seen examples of how it was done. @kianenigma above suggested to use type Executive, I am not sure it would achieve the same, and I have not seen how it could be used in the context of migrations, that's why I still relied on that trait, which is used by a lot of other pallets that implemented migrations from V2toV3.

Reply:
ah I see, substrate v2 and susbtrate v3.

Yes I personally like to have this additional trait to put the information needed for the migration function, this make the migration module completely isolated.
But I'm not sure if everybody like them.

But the PR looks good except from the few grumbles below.